File: LeetCode-Python/BinaryTreeZigzagLevelOrderTraversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root: 
            return [] 
        result = [] 
        nodeDeque = deque([root]) 
        leftToRight = True 
        while nodeDeque: 
            level = [] 
            size = len(nodeDeque) 
            for _ in range(size): 
                if leftToRight: 
                    node = nodeDeque.popleft() 
                    level.append(node.val) 
                    if node.left: 
                        nodeDeque.append(node.left) 
                    if node.right: 
                        nodeDeque.append(node.right)
                else: 
                    node = nodeDeque.pop() 
                    level.append(node.val) 
                    if node.right: 
                        nodeDeque.appendleft(node.right) 
                    if node.left: 
                        nodeDeque.appendleft(node.left) 
            result.append(level)
            leftToRight = not leftToRight 
        return result

File: LeetCode-Cpp/BinaryTreeZigzagLevelOrderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> result; 
        if (!root) return result; 
        deque<TreeNode*> nodeDeque; 
        nodeDeque.push_back(root); 
        bool leftToRight = true; 
        while (!nodeDeque.empty()) { 
            int size = nodeDeque.size(); 
            vector<int> level(size); 
            for (int i = 0; i < size; ++i) { 
                TreeNode* node; 
                if (leftToRight) { 
                    node = nodeDeque.front(); 
                    nodeDeque.pop_front(); 
                    if (node->left) nodeDeque.push_back(node->left); 
                    if (node->right) nodeDeque.push_back(node->right); 
                } 
                else { 
                        node = nodeDeque.back(); 
                        nodeDeque.pop_back(); 
                        if (node->right) nodeDeque.push_front(node->right); 
                        if (node->left) nodeDeque.push_front(node->left); } level[i] = node->val; 
                        } 
                        result.push_back(level); 
                        leftToRight = !leftToRight; 
                    } 
                return result;
    }
};
